cmd/zar/zq/command.go

```diff
@@ -0,0 +1,238 @@
+package zq
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/brimsec/zq/archive"
+	"github.com/brimsec/zq/ast"
+	"github.com/brimsec/zq/cmd/zar/root"
+	"github.com/brimsec/zq/driver"
+	"github.com/brimsec/zq/emitter"
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/scanner"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio"
+	"github.com/brimsec/zq/zio/detector"
+	"github.com/brimsec/zq/zio/ndjsonio"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zql"
+	"github.com/mccanne/charm"
+	"go.uber.org/zap"
+)
+
+var Zq = &charm.Spec{
+	Name:  "zq",
+	Usage: "zq [-R dir] [options]",
+	Short: "walk an archive and run zq logic",
+	Long: `
+"zar zq" descends the directory given by the -R option (or ZAR_ROOT env) looking for
+logs with zar directories and for each such directory found, it runs
+the zq logic relative to that directory.  The file names here are relative to that
+directory and the special name "_" refers to the actual log file
+in the parent of the zar directory.
+
+If the root directory is not specified by either the ZAR_ROOT environemnt
+variable or the -R option, then the current directory is assumed.
+`,
+	New: New,
+}
+
+func init() {
+	root.Zar.Add(Zq)
+}
+
+type Command struct {
+	*root.Command
+	root           string
+	jsonTypePath   string
+	jsonPathRegexp string
+	jsonTypeConfig *ndjsonio.TypeConfig
+	outputFile     string
+	stopErr        bool
+	quiet          bool
+	ReaderFlags    zio.ReaderFlags
+	WriterFlags    zio.WriterFlags
+}
+
+func fileExists(path string) bool {
+	if path == "-" {
+		return true
+	}
+	info, err := os.Stat(path)
+	if err != nil {
+		return false
+	}
+	return !info.IsDir()
+}
+
+func New(parent charm.Command, f *flag.FlagSet) (charm.Command, error) {
+	c := &Command{Command: parent.(*root.Command)}
+	f.StringVar(&c.root, "R", os.Getenv("ZAR_ROOT"), "root directory of zar archive to walk")
+	f.BoolVar(&c.quiet, "q", false, "don't display zql warnings")
+	// XXX f.StringVar(&c.dir, "d", "", "directory for output data files")
+	f.StringVar(&c.outputFile, "o", "", "write data to output file")
+	f.StringVar(&c.jsonTypePath, "j", "", "path to json types file")
+	f.StringVar(&c.jsonPathRegexp, "pathregexp", c.jsonPathRegexp, "regexp for extracting _path from json log name (when -inferpath=true)")
+	// XXX f.BoolVar(&c.quiet, "q", false, "don't display zql warnings")
+	f.BoolVar(&c.stopErr, "e", true, "stop upon input errors")
+
+	// Flags added for writers are -f, -T, -F, -E, -U, and -b
+	c.WriterFlags.SetFlags(f)
+
+	// Flags added for readers are -i XXX json
+	c.ReaderFlags.SetFlags(f)
+
+	return c, nil
+}
+
+func (c *Command) loadJsonTypes() (*ndjsonio.TypeConfig, error) {
```

Comment:
can we get rid of all this json stuff here? or do that as part of the "xxx refactor into a tools package" comment?

Reply:
I would rather leave it as a placeholder so we remember to adjust this call site.  The log files here referred to with _ can be json in need of schema maps.